code/modules/plumbing/ducts.dm

```diff
@@ -132,12 +137,27 @@ All the important duct code:
 		else
 			create_duct()
 			duct.add_duct(D)
+
 	add_neighbour(D, direction)
-	//tell our buddy its time to pass on the torch of connecting to pipes. This shouldn't ever infinitely loop since it only works on pipes that havent been inductrinated
-	D.attempt_connect()
+	D.add_neighbour(src, opposite_dir)
+	D.add_connects(opposite_dir)
+
+	D.safe_connect(FALSE)
 
 	return TRUE
 
+///Delegates connecting to the plumbing subsystem, which ensures connections are done sane and clean
+/obj/machinery/duct/proc/safe_connect(prime = TRUE)
+
+	//safe option, waits for all other ducts to finish and then politely asks if it could perhaps be connected, m'lady
+	if(prime)
+		SSfluids.prime_queued += src
+		SSfluids.attempt_quick_build()
+	else //this should only be done through other ducts
+		//addtimer so it's delegated to a proper queued subsystem. might cause large networks of ducts to visually connect and reconnect, but prevents lagspikes and crashes
+		SSfluids.build_queued += src //track we're already being built
+		addtimer(CALLBACK(src, .proc/attempt_connect))
```

Comment:
I'm not quite sure why you're using a timer here, even with that explanation. Could you explain what it prevents exactly?

Reply:
A duct used to, on Initialize, check for all other ducts and then connect to them. They would recursively ping each other untill the whole network was constructed. When a bunch of the same function is called from the same type of object through recursion, byond kills it when it hits the 60~ ducts ballpark with its recursion prevention.

Addtimer means they’re called one-by-one instead of a giant recursive network thats all loaded in the same tick. Only one duct is constructed per worldtick, instead of all at once, which doesn’t trigger byonds infinite loop prevention. The subsystem delegation is purely to accomodate this change so we don’t have 20 neighbouring ducts asynchronously trying to make one network and slowing the game to a halt, which is what happened on my end

Reply:
Why use the timer subsystem for this? wouldn't the fluids subsystem be better suited to it?

Reply:
Fluids subsystem mostly handles routine plumbing stuff, and once every two seconds. I’d need to make a seperate subsystem to accommodate it. 

I’m gonna try it without the subsystem since I’ve done a lot of tweaks along the way, it might just work without it now